proto/pulumi/resource_status.proto

```diff
@@ -0,0 +1,64 @@
+// Copyright 2025, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+import "google/protobuf/struct.proto";
+import "pulumi/provider.proto";
+
+package pulumirpc;
+
+option go_package = "github.com/pulumi/pulumi/sdk/v3/proto/go;pulumirpc";
+
+// ResourceStatus is an interface that can be called from a resource provider to update status about a resource.
+service ResourceStatus {
+    // `PublishViewSteps` is used to publish a series of steps for a view resource.
+    // Views can be materialized via create and update steps, and more complex
+    // changes, such as replacements, can be modeled as a series of steps.
+    // The engine does not actually apply these steps, but rather flows them through
+    // the engine such that the view resources are written to state and the view
+    // resources are displayed in the UI.
+    rpc PublishViewSteps(PublishViewStepsRequest) returns (PublishViewStepsResponse) {}
+}
+
+// `PublishViewStepsRequest` is the type of requests sent as part of a
+// [](pulumirpc.ResourceStatus.PublishViewSteps) call.
+message PublishViewStepsRequest {
+    string token = 1; // the service context token
+    string type = 2; // the type of the view resource
+    string name = 3; // the name, for URN purposes, of the view resource
+    repeated ViewStep steps = 4; // the steps to publish
+}
+
+// `PublishViewStepsResponse` is the type of responses sent as part of a
+// [](pulumirpc.ResourceStatus.PublishViewSteps) call.
+message PublishViewStepsResponse {
+}
+
+// `ViewStep` represents a deployment operation step for a view resource.
+message ViewStep {
+    string op = 1; // the operation being performed
+    ViewStepState old = 2; // the state of the view resource before performing the step
+    ViewStepState new = 3; // the state of the view resource after performing the step
+    repeated string keys = 4; // the keys causing a replacement (only applicable for "create" and "replace" ops)
+    repeated string diffs = 5; // the keys that changed with this step
+    map<string, PropertyDiff> detailed_diff = 6;
+    bool has_detailed_diff = 7; // whether the detailed diff is present
```

Comment:
Can we mark detailed_diff as optional and then use protobuf presence checks rather than a separate `has_` field?

Reply:
`optional` isn't supported on maps, unfortunately. We'd have to use some other workaround, like this `has_` field, or wrap it in another message, or use oneof or something like that.

I've just one what we've done in `DiffResponse`:

https://github.com/pulumi/pulumi/blob/558f3bb70c29e60d6b829be55e493e5bc1fc7f4d/proto/pulumi/provider.proto#L742-L749